### PR TITLE
Test and officially bless generic views for use by plugins

### PR DESCRIPTION
**Pull request by jathanism**
### Fixes: #14 

Moved `dummy_plugin` to `examples/dummy_plugin` at project root

- Added a `README.md` and a `pyproject.toml` for it
- Added it as a dev dependency to nautobot's root `pyproject.toml`
- Add a version string to the plugin
- Added a `to_csv()` method to `DummyModel`
- Asserted that unit tests ran
- Implemented generic views for DummyModel
  - Added URLs and everything
  - Added a detail view template
  - Added tables, filters, forms, search, changelog
- Updated plugin code quite a bit to use correct bases for API serializers/views to get all tests passing
- Changed base of API `DummyModelViewSet` to use `nautobot.core.api.views.ModelViewSet`
- Changed base of API `DummyModelSerializer` to use `nautobot.core.api.serializers.ValidatedModelSerializer`
- Added `url` field to API `DummyModelSerializer`
- Switched API urls to use `nautobot.core.api.OrderedDefaultRouter` to get bulk update/delete working correctly.
- Updated Invoke tasks to also run black/flake8 on `examples/`
- Added a `poetry.lock` file for `dummy_plugin`

### CODE REVIEW

File: examples/dummy_plugin/dummy_plugin/__init__.py [hunk 1]
```diff
@@ -2,14 +2,14 @@
...
+    name = "dummy_plugin"
     verbose_name = "Dummy plugin"
     version = "0.0"
     description = "For testing purposes only"
     base_url = "dummy-plugin"
     min_version = "0.9"
     max_version = "9.0"
-    middleware = ["nautobot.extras.tests.dummy_plugin.middleware.DummyMiddleware"]
+    middleware = ["dummy_plugin.middleware.DummyMiddleware"]
     installed_apps = ["nautobot.extras.tests.dummy_plugin_dependency"]
```
Comment (glennmatthews) [inline examples/dummy_plugin/dummy_plugin/__init__.py:13]: Is the installed_app going to get moved to `examples` as well?

File: examples/dummy_plugin/dummy_plugin/__init__.py [hunk 2]
```diff
@@ -2,14 +2,14 @@
 
 
 class DummyPluginConfig(PluginConfig):
-    name = "nautobot.extras.tests.dummy_plugin"
+    name = "dummy_plugin"
     verbose_name = "Dummy plugin"
     version = "0.0"
```
Comment (glennmatthews) [inline examples/dummy_plugin/dummy_plugin/__init__.py:7]: I see that `pyproject.toml` declares the version as 1.0.0 - would be good to have them match, perhaps using:

```python

try:
    from importlib import metadata
except ImportError:
    # Python version < 3.8
    import importlib_metadata as metadata

__version__ = metadata.version(__name__)

...

    version = __version__
```

**Review by glennmatthews — COMMENTED** [review]
This looks like a great start!

File: examples/dummy_plugin/dummy_plugin/__init__.py (hunk 2 above)
Comment (jathanism) [reply examples/dummy_plugin/dummy_plugin/__init__.py:7]: Yes.

File: examples/dummy_plugin/dummy_plugin/__init__.py (hunk 1 above)
Comment (jathanism) [reply examples/dummy_plugin/dummy_plugin/__init__.py:13]: I am on the fence about the dummy dependency moving.

File: examples/dummy_plugin/dummy_plugin/api/serializers.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+from dummy_plugin.models import DummyModel
+
+
+class DummySerializer(ValidatedModelSerializer):
+    url = serializers.HyperlinkedIdentityField(view_name="plugins-api:dummy_plugin-api:dummymodel-detail")
+
+    class Meta:
+        model = DummyModel
+        fields = ("url", "id", "name", "number")
```
Comment (glennmatthews) [inline examples/dummy_plugin/dummy_plugin/api/serializers.py:13]: Not blocking for this PR but at some point for completeness we may want to add a NestedDummySerializer as well.

File: examples/dummy_plugin/dummy_plugin/forms.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    """Filtering/search form for `DummyModel` objects."""
+
+    model = DummyModel
+    q = forms.CharField(required=False, label="Search")
+    name = forms.CharField(max_length=20, required=False)
+    number = forms.IntegerField(required=False)
+
+
+class DummyModelBulkEditForm(BootstrapMixin, BulkEditForm):
+    """Bulk edit/delete form for `DummyModel` objects."""
```
Comment (glennmatthews) [inline examples/dummy_plugin/dummy_plugin/forms.py:38]: I don't think the `BulkDeleteView` actually uses this form?

File: examples/dummy_plugin/dummy_plugin/templates/dummy_plugin/dummymodel.html [hunk 5]
```diff
@@ -0,0 +1,76 @@
+{% extends 'base.html' %}
+{% load buttons %}
+{% load custom_links %}
+{% load helpers %}
+{% load plugins %}
+{% load static %}
+{% load tz %}
```
Comment (glennmatthews) [inline examples/dummy_plugin/dummy_plugin/templates/dummy_plugin/dummymodel.html:7]: Are `static` and `tz` actually needed here? I'm not generally loading them in my plugin "detail" templates.

File: examples/dummy_plugin/dummy_plugin/templates/dummy_plugin/dummymodel.html [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    <ul class="nav nav-tabs">
+        <li role="presentation"{% if not active_tab %} class="active"{% endif %}>
+            <a href="{{ object.get_absolute_url }}">Dummy Model</a>
+        </li>
+        {% if perms.extras.view_objectchange %}
+            <li role="presentation"{% if active_tab == 'changelog' %} class="active"{% endif %}>
+                <a href="{% url 'plugins:dummy_plugin:dummymodel_changelog' pk=object.pk %}">Change Log</a>
+            </li>
+        {% endif %}
+    </ul>
```
Comment (glennmatthews) [inline examples/dummy_plugin/dummy_plugin/templates/dummy_plugin/dummymodel.html:49]: This block is such total boilerplate - I wish we had (overridable if necessary) stuff in the generic template for this.

File: nautobot/extras/tests/test_plugins.py [hunk 7]
```diff
@@ -262,24 +253,91 @@ class DummyConfigWithInstalledApps(dummy_config):
...
-class PluginAPITestCase(APITestCase):
-    @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"])
-    def test_api_views(self):
+class PluginGenericViewTestCase(
+    ViewTestCases.CreateObjectViewTestCase,
+    ViewTestCases.DeleteObjectViewTestCase,
+    ViewTestCases.EditObjectViewTestCase,
+    ViewTestCases.GetObjectViewTestCase,
+    ViewTestCases.GetObjectChangelogViewTestCase,
+    ViewTestCases.ListObjectsViewTestCase,
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_plugins.py:265]: Add `BulkImport`, `BulkEdit`, `BulkDelete` test mixins as well?

File: examples/dummy_plugin/dummy_plugin/forms.py (hunk 4 above)
Comment (jathanism) [reply examples/dummy_plugin/dummy_plugin/forms.py:38]: Ah yeah that was just copypasta.

File: examples/dummy_plugin/dummy_plugin/templates/dummy_plugin/dummymodel.html (hunk 5 above)
Comment (jathanism) [reply examples/dummy_plugin/dummy_plugin/templates/dummy_plugin/dummymodel.html:7]: Just more copypasta. I can remove it for brevity.

File: examples/dummy_plugin/dummy_plugin/templates/dummy_plugin/dummymodel.html (hunk 6 above)
Comment (jathanism) [reply examples/dummy_plugin/dummy_plugin/templates/dummy_plugin/dummymodel.html:49]: Yeah like a generic detail view base or something? I'm there for it.

Comment (glennmatthews) [reply examples/dummy_plugin/dummy_plugin/templates/dummy_plugin/dummymodel.html:49]: https://github.com/nautobot/nautobot/issues/479

File: nautobot/extras/tests/test_plugins.py (hunk 7 above)
Comment (jathanism) [reply nautobot/extras/tests/test_plugins.py:265]: Hacknowledged.

**Review by glennmatthews — APPROVED** [review]
Nice work! Would it make sense to add anything in `nautobot/docs` that points to the dummy plugin? Maybe not since the docs are hosted separately from the repository, but wanted to see what you think.

**Comment by jathanism** [comment]
> Nice work! Would it make sense to add anything in `nautobot/docs` that points to the dummy plugin? Maybe not since the docs are hosted separately from the repository, but wanted to see what you think.

I think it would make sense to include such documentation along side the plugin development docs maybe?

**Pull request merged**